static/app/utils/profiling/flamegraph/selectNearestFrame.ts

```diff
@@ -0,0 +1,147 @@
+ * FlamegraphFrame relative to the starting node.
+ *
+ * below is a gist of the walking algorithm:
+ *
+ *  up    =>
+ *          |> if parent is present, goto parent
+ *          |> if no parent & previousSibling is present, goto maxDepth of previousSibling
+ *          |> if no parent & no previousSibling, no-op
+ *  down  =>
+ *          |> if immediate child is present, goto first child
+ *          |> if no child & walk up to nearest nextSibling
+ *          |> if at max depth & no nextSibling, no-op
+ *  left  =>
+ *          |> walk up to find a node that has a previousSibling
+ *          |> walk down the sibling tree following right branches
+ *          |> return node matching target depth, if max depth found is less than target return
+ *  right => same as left, but we find the right sibling, and follow left branches down until target/max depth is reached
+ */
+export function selectNearestFrame(frame: FlamegraphFrame, direction: Direction) {
+  const targetDepth = frame.depth;
+  const parent = frame.parent;
+  const node = frame;
+
+  if (direction === 'up') {
+    if (parent) {
+      return parent;
+    }
+
+    const leftSibling = getSibling(node, 'left');
+    if (leftSibling) {
+      return leftSibling;
+    }
+
+    return frame;
+  }
+
+  const child = frame.children?.[0];
+  if (direction === 'down') {
+    if (child) {
+      return child;
+    }
+    const sibling = scanForNearestSibling(node, 'right');
+    if (sibling) {
+      return sibling;
+    }
+
+    return frame;
+  }
+
+  const sibling = scanForNearestSibling(node, direction);
+
+  if (sibling) {
+    const nodeWithDepth = scanForNearestFrameWithDepth(sibling, targetDepth, direction);
+    if (nodeWithDepth) {
+      return nodeWithDepth;
+    }
+  }
+
+  return frame;
+}
+
+function scanForNearestSibling(frame: FlamegraphFrame, directionX: DirectionX) {
+  let node = frame;
+  while (node) {
+    let parent = node.parent;
+    const sibling = getSibling(node, directionX);
+
+    if (sibling) {
+      return sibling;
+    }
+
+    if (!parent) {
+      break;
+    }
+
+    node = parent;
+    parent = parent.parent;
+  }
+  return frame;
+}
+
+/**
+ *  scanForNearestFrameWithDepth will walk down a FlamegraphFrame looking for a target depth.
+ *  it will follow either the furthest left or right branch based on direction.
+ *  if target depth is not found, we return the frame with the max depth along that branch.
+ */
+function scanForNearestFrameWithDepth(
+  frame: FlamegraphFrame,
+  depth: number,
+  directionX: DirectionX
+) {
+  let node = frame;
+
+  while (node) {
+    if (node.depth === depth) {
+      return node;
+    }
+    if (!node.children) {
+      continue;
+    }
+
+    const nextNode = node.children[directionX === 'right' ? 0 : node.children.length - 1];
+    if (!nextNode) {
+      return node;
+    }
+    node = nextNode;
+  }
+
+  return frame;
```

Comment:
Same return comment here

Reply:
technically this return is unreachable, but the linter loses its shit.

Reply:
Ok, then I would propose to break instead of return on L103 and return node. This way the while loop only does the searching. If I'm not mistaken, I think you could rewrite the while loop as something like 
```
while(node && node.depth !== depth){
  node = node.children[directionX === 'right' ? 0 : node.children.length - 1];
}
return node
```

Reply:
That won't work. `node` can possibly be undefined in this case, breaking its behavior. 

the idea is to find the target depth OR if you've reached the bottom of the branch return that node.

I'd prefer not to change this as i find it easy to follow as is.